File: Phase01/src/InvertedIndexSearch.java
package Phase01.src;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class InvertedIndexSearch {

    private HashMap<String, ArrayList<Entry>> indexMap = new HashMap<String, ArrayList<Entry>>();
    private HashSet<Entry> allDocs = new HashSet<Entry>();
    private static final String NORMAL_PATTERN = "\\s(\\w+)";
    private static final String PLUS_PATTERN = "\\+(\\w+)";
    private static final String MINUS_PATTERN = "-(\\w+)";
    private static final String NOT_WORDS_PATTERN = "[^a-z0-9]";
    private static final String SPACE_PATTERN = "\\s+";

    public InvertedIndexSearch(HashMap<String, String> fileContents) {
        for (String id : fileContents.keySet()) {
            addWords(id, getListOfWords(fileContents.get(id)));
            allDocs.add(new Entry(id, 0));
        }
    }

    public HashMap<String, ArrayList<Entry>> getIndexMap(){
        return this.indexMap;
    }

    public Result search(String input) {
        Boolean multiWords = input.contains(" ") || input.contains("+") || input.contains("-");
        return new Result(this.searchQuery(input), multiWords);
    }

    public HashSet<Entry> searchQuery(String query) {

        var normalSet = new HashSet<Entry>(allDocs);
        var plusSet = new HashSet<Entry>();
        var minusSet = new HashSet<Entry>();
        for (String normalStr : this.seperatQuery(" " + query, NORMAL_PATTERN)) {
            var temp = searchWord(normalStr);
            temp.retainAll(normalSet);
            normalSet = temp;
        }
        for (String plusStr : this.seperatQuery(query, PLUS_PATTERN))
            plusSet.addAll(searchWord(plusStr));
        for (String minusStr : this.seperatQuery(query, MINUS_PATTERN))
            minusSet.addAll(searchWord(minusStr));
        return this.combineResults(normalSet, plusSet, minusSet);
    }

    public HashSet<Entry> searchWord(String word) {
        var result = new HashSet<Entry>();
        String word_s = "";
        if (word.charAt(word.length() - 1) == 's') {
            word_s = word.substring(0, word.length() - 1);
        } else {
            word_s = word + "s";
        }
        if (indexMap.containsKey(word)) {
            result.addAll(indexMap.get(word));
        }
        if (indexMap.containsKey(word_s)) {
            result.addAll(indexMap.get(word_s));
        }
        return result;
    }

    public HashSet<Entry> combineResults(HashSet<Entry> normalSet, HashSet<Entry> plusSet, HashSet<Entry> minusSet) {
        // add normal set, Intersect plus set, Minus minus set'
        var result = new HashSet<Entry>();
        result.addAll(normalSet);
        result.addAll(plusSet);
        result.removeAll(minusSet);
        return result;
    }

    public ArrayList<String> seperatQuery(String query, String regex) {
        var words = new ArrayList<String>();
        Pattern pattern = Pattern.compile(regex);
        Matcher matcher = pattern.matcher(query);
        while (matcher.find()) {
            words.add(matcher.group(1));
        }
        return words;
    }

    public void addWords(String id, String[] listOfWords) {
        for (int i = 0; i < listOfWords.length; i++) {
            String word = listOfWords[i];
            if (StopWordsClass.STOP_WORDS.contains(word))
                continue;
            if (indexMap.containsKey(word)) {
                indexMap.get(word).add(new Entry(id, i + 1));
            } else {
                indexMap.put(word, new ArrayList<Entry>(Arrays.asList(new Entry(id, i + 1))));
            }
        }
    }

    public String[] getListOfWords(String content) {
        content = content.toLowerCase().replaceAll(NOT_WORDS_PATTERN, " ").trim();
        return content.split(SPACE_PATTERN);
    }

}

File: Phase05/app/library/InvertedIndexSearch.cs
using System.IO;
using System.Collections;
using System.Text.RegularExpressions;
using System.Collections.Generic;

namespace library
{
    public class InvertedIndexSearch
    {
        public Dictionary<string, List<Entry>> IndexMap { get; set; } = new Dictionary<string, List<Entry>>();
        public HashSet<Entry> allDocs { get; set; } = new HashSet<Entry>();
        public static string SPACE_PATTERN1 => SPACE_PATTERN;
        public const string NORMAL_PATTERN = @"\s(\w+)";
        public const string PLUS_PATTERN = @"\+(\w+)";
        public const string MINUS_PATTERN = @"-(\w+)";
        private const string NOT_WORDS_PATTERN = @"[^a-z0-9]";
        private const string SPACE_PATTERN = @"\s+";

        public InvertedIndexSearch(Dictionary<string, string> fileContents)
        {
            foreach (var (id, contnet) in fileContents)
            {
                AddWords(id, GetListOfWords(contnet));
                allDocs.Add(new Entry(id, 0));
            }
        }

        public HashSet<Entry> CombineResults(HashSet<Entry> normalSet, HashSet<Entry> plusSet, HashSet<Entry> minusSet)
        {
            var result = new HashSet<Entry>();
            result.UnionWith(normalSet);
            result.UnionWith(plusSet);
            result.ExceptWith(minusSet);
            return result;
        }

        public List<string> SeperateQuery(string query, string pattern)
        {
            var words = new List<string>();
            Regex regex = new Regex(pattern);
            MatchCollection mathces = regex.Matches(query);
            foreach (Match match in mathces)
                words.Add(match.Groups[1].Value);
            return words;
        }

        public void AddWords(string id, string[] listOfWords)
        {
            for (int i = 0; i < listOfWords.Length; i++)
            {
                string word = listOfWords[i];
                if (StopWordsClass.STOP_WORDS.Contains(word))
                    continue;
                if (IndexMap.ContainsKey(word))
                    IndexMap[word].Add(new Entry(id, i + 1));
                else
                    IndexMap.Add(word, new List<Entry> { new Entry(id, i + 1) });
            }
        }
        public string[] GetListOfWords(string content)
        {
            content = Regex.Replace(content.ToLower(), NOT_WORDS_PATTERN, " ").Trim();
            return Regex.Split(content, SPACE_PATTERN);
        }
        public HashSet<Entry> SearchWord(string word)
        {
            var result = new HashSet<Entry>();
            string word_s;
            if (word.EndsWith("s"))
                word_s = word.Remove(word.Length - 1);
            else
                word_s = word + "s";
            if (IndexMap.ContainsKey(word))
                result.UnionWith(IndexMap[word]);
            if (IndexMap.ContainsKey(word_s))
                result.UnionWith(IndexMap[word_s]);
            return result;
        }
        public HashSet<Entry> SearchQuery(string query)
        {
            var normalSet = new HashSet<Entry>(allDocs);
            var plusSet = new HashSet<Entry>();
            var minusSet = new HashSet<Entry>();
            foreach (string normalStr in this.SeperateQuery(" " + query, NORMAL_PATTERN))
            {
                var temp = SearchWord(normalStr);
                temp.IntersectWith(normalSet);
                normalSet = temp;
            }
            foreach (string plusStr in this.SeperateQuery(query, PLUS_PATTERN))
                plusSet.UnionWith(SearchWord(plusStr));
            foreach (string minusStr in this.SeperateQuery(query, MINUS_PATTERN))
                minusSet.UnionWith(SearchWord(minusStr));
            return this.CombineResults(normalSet, plusSet, minusSet);
        }

        public Result Search(string input)
        {
            var multiWords = input.Contains(" ") || input.Contains("+") || input.Contains("-");
            return new Result(this.SearchQuery(input), multiWords);
        }
    }
}